### PR TITLE
Support header in Routing Form CSV

### LINKED ISSUE #3948: [CAL-250] routing forms .csv download needs a header
when clicking "download csv"
CleanShot 2022-08-24 at 18 56 24@2x

it'll give you a csv with this format:
CleanShot 2022-08-24 at 18 56 07@2x

whereas we need a format like this:

<table>
<tr>
<td> What Problem are you trying to solve?
<td> Do you plan to Self-Host?
<td> ..
<td> ..
<tr>
<td> I want to schedule sales calls
<td> True
<td> ..
<td> ..
<tr>
<td> Workout with a trainer
<td> True
<td> ..
<td> ..
</table>

the way to achieve this is just have the top row be the column name

> From [Linear-GitHub Sync](https://synclinear.com)

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

- Fixes #3948 by doing soft deletion of fields instead of hard deletion. This wasn't possible earlier because of changing number of fields(due to addition and deletion of fields) and thus changing number of columns at different moments in time. But if we stop deleting fields, then columns would only increase and can never decrease.
- Also adds a new column(Submission Time)

Sample CSV - Deleted fields would come after all fields of form
Screenshot 2022-10-25 at 3 28 53 PM

**Environment**:  Production

## Type of change

- [x] Improvement

## How should this be tested?

- [x] Create a Form and add 3 fields(F1,F2,F3). Submit the form with all 3 fields. Delete the last field and check CSV now. It should have last field with it's data still there.
- [x] Now, to the previous form add a new field(there are still 3 fields as per user(F1, F2, F4)). Fill the form and check CSV now. It should show F3 as empty for second response.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: packages/app-store/ee/routing-forms/api/responses/[formId].ts [hunk 1]
```diff
@@ -22,25 +29,22 @@ async function* getResponses(formId: string) {
...
           serializedValue = value.map((val) => escapeCsvText(val)).join(" | ");
         } else {
           serializedValue = escapeCsvText(value);
         }
-        csvLineColumns.push(`"${label} :=> ${serializedValue}"`);
-      }
-      csv.push(csvLineColumns.join(","));
+        csvCells.push(serializedValue);
+      });
+      csvCells.push(response.createdAt.toISOString());
```
Comment (hariombalhara) [inline packages/app-store/ee/routing-forms/api/responses/[formId].ts:46]: Push new `Submission Time` column value

File: packages/app-store/ee/routing-forms/api/responses/[formId].ts [hunk 2]
```diff
@@ -63,13 +67,28 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
...
   res.setHeader("Transfer-Encoding", "chunked");
-  const csvIterator = getResponses(formId);
+  const headerFields = serializableForm.fields || [];
+  const csvIterator = getResponses(formId, headerFields);
+
+  // Make Header
+  res.write(
+    headerFields
+      .map((field) => `${field.label}${field.deleted ? "(Deleted)" : ""}`)
+      .concat(["Submission Time"])
```
Comment (hariombalhara) [inline packages/app-store/ee/routing-forms/api/responses/[formId].ts:88]: Mark deleted fields as deleted and add new header column for Submission Time

File: packages/app-store/ee/routing-forms/lib/getSerializableForm.ts [hunk 3]
```diff
@@ -17,7 +23,11 @@ export function getSerializableForm<TForm extends App_RoutingForms_Form>(form: T
   // Ideally we shouldb't have needed to explicitly type it but due to some reason it's not working reliably with VSCode TypeCheck
   const serializableForm: SerializableForm<TForm> = {
     ...form,
-    fields: fieldsParsed.data,
+    fields: fieldsParsed.data
```
Comment (hariombalhara) [inline packages/app-store/ee/routing-forms/lib/getSerializableForm.ts:26]: Centrally support not sending soft deleted fields. Fields can only be soft deleted.

File: packages/app-store/ee/routing-forms/lib/getSerializableForm.ts [hunk 4]
```diff
@@ -3,7 +3,13 @@ import { App_RoutingForms_Form } from "@prisma/client";
 import { SerializableForm } from "../types/types";
 import { zodFields, zodRoutes } from "../zod";
 
-export function getSerializableForm<TForm extends App_RoutingForms_Form>(form: TForm) {
+/**
+ * Doesn't have deleted fields by default
+ */
+export function getSerializableForm<TForm extends App_RoutingForms_Form>(
+  form: TForm,
+  withDeletedFields = false
```
Comment (hariombalhara) [inline packages/app-store/ee/routing-forms/lib/getSerializableForm.ts:11]: Allow retrieving deleted fields

File: packages/app-store/ee/routing-forms/playwright/tests/basic.e2e.ts [hunk 5]
```diff
@@ -215,13 +215,22 @@ test.describe("Routing Forms", () => {
...
-        `
-"Test field :=> event-routing"
-"Test field :=> external-redirect"
-"Test field :=> custom-page"`.trim()
-      );
+      const firstResponseCells = csvRows[1].split(",");
+      const secondResponseCells = csvRows[2].split(",");
+      const thirdResponseCells = csvRows[3].split(",");
+
+      expect(firstResponseCells.slice(0, -1).join(",")).toEqual("event-routing,");
```
Comment (hariombalhara) [inline packages/app-store/ee/routing-forms/playwright/tests/basic.e2e.ts:226]: Can't verify entire CSV file at once. Parse line by line.

File: packages/app-store/ee/routing-forms/playwright/tests/basic.e2e.ts [hunk 6]
```diff
@@ -215,13 +215,22 @@ test.describe("Routing Forms", () => {
...
+      expect(new Date(firstResponseCells.at(-1)).getDay()).toEqual(new Date().getDay());
```
Comment (hariombalhara) [inline packages/app-store/ee/routing-forms/playwright/tests/basic.e2e.ts:227]: Can't match exact time of form submission(as it is variable). Check just the date for now.

File: packages/app-store/ee/routing-forms/trpc-router.ts [hunk 7]
```diff
@@ -41,15 +41,9 @@ const app_RoutingForms = createRouter()
...
             });
           }
-          const fieldsParsed = zodFields.safeParse(form.fields);
-          if (!fieldsParsed.success) {
-            // This should not be possible normally as before saving the form it is verified by zod
-            throw new TRPCError({
-              code: "INTERNAL_SERVER_ERROR",
-            });
-          }
+          const parsedForm = getSerializableForm(form);
```
Comment (hariombalhara) [inline packages/app-store/ee/routing-forms/trpc-router.ts:44]: getSerializedForm already handles fieldsParsed failure case.

File: packages/app-store/ee/routing-forms/trpc-router.ts [hunk 8]
```diff
@@ -41,15 +41,9 @@ const app_RoutingForms = createRouter()
...
 
-          const fields = fieldsParsed.data;
+          const fields = parsedForm.fields;
```
Comment (hariombalhara) [inline packages/app-store/ee/routing-forms/trpc-router.ts:46]: Automatically retrieves non-deleted fields only.

**Review by hariombalhara — COMMENTED** [review]
Self review done

**Comment by hariombalhara** [comment]
cc @CarinaWolli. Now that you are contributing to Routing Forms, I have someone to better review my Routing Forms PR.

**Review by CarinaWolli — APPROVED** [review]
Works as expected and code looks good 👏🏻

**Pull request merged**